tracing-subscriber/src/filter/targets.rs

```diff
@@ -344,20 +379,111 @@ impl<S> layer::Filter<S> for Targets {
     }
 }
 
+impl IntoIterator for Targets {
+    type Item = (String, LevelFilter);
+
+    type IntoIter = IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        fn unpack(directive: StaticDirective) -> Option<(String, LevelFilter)> {
+            let level = directive.level;
+            directive.target.map(move |target| (target, level))
+        }
+        IntoIter(self.0.into_iter().filter_map(unpack))
+    }
+}
+
+impl<'a> IntoIterator for &'a Targets {
+    type Item = (&'a str, LevelFilter);
+
+    type IntoIter = Iter<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        fn unpack(directive: &StaticDirective) -> Option<(&str, LevelFilter)> {
+            directive
+                .target
+                .as_deref()
+                .map(|target| (target, directive.level))
+        }
+        Iter(self.0.iter().filter_map(unpack))
+    }
+}
+
+/// An owning iterator over the [target]-[level] pairs of a `Targets` filter.
+///
+/// This struct is created by the `IntoIterator` trait implementation of [`Targets`].
+///
+/// # Examples
+///
+/// Merge the targets from one `Targets` with another:
+///
+/// ```
+/// use tracing_subscriber::filter::Targets;
+/// use tracing_core::Level;
+///
+/// let mut filter = Targets::new().with_target("my_crate", Level::INFO);
+/// let overrides = Targets::new().with_target("my_crate::interesting_module", Level::DEBUG);
+///
+/// filter.extend(overrides);
+/// # drop(filter);
+/// ```
+///
+/// [target]: tracing_core::Metadata::target
+/// [level]: tracing_core::Level
+#[derive(Debug)]
+pub struct IntoIter(
+    #[allow(clippy::type_complexity)] // alias indirection would probably make this more confusing
+    std::iter::FilterMap<
+        <DirectiveSet<StaticDirective> as IntoIterator>::IntoIter,
```

Comment:
we might be able to avoid the type complexity lint if we just made a type alias in `directives.rs` that's either `std::vec::IntoIter`or `smallvec::IntoIter`...but it may not really matter, this looks fine as-is!

Reply:
I did actually start with that, but then decided introducing another alias might be confusing that just referring to the trait. But perhaps I was too stingy? I'll leave it for now but would change it if you do prefer.

Reply:
:woman_shrugging: i'm fine either way --- this seems good to me!